### ISSUE TITLE
Standarize player ship indexes

### ISSUE THREAD

**Issue by IvanSanchez** [artemis-nerds/node-asbs-lib#11]
The player ship indexes have been bugging me for some time. In some packets (e.g. console status, setPlayerShipIndex) the field is 0-indexed; while in some other packets (player ship updates) is 1-indexed. So sometimes the player ship index goes 0-7 and sometimes 1-8.

_Maybe_ it would be a good idea to standarize these into the same, by creating an offset integer datatype. **_Maybe**_.

**Comment by mrfishie** [comment]
We could actually define an 'in-place' type so that we aren't cluttering the types file with ones that are only going to be used once.

For example:

``` js
... new struct({
        playerShipIndex: {
            pack: function(buffer, value) {
                type.int32.pack(buffer, value - 1);
            },
            unpack: function(buffer) {
                return type.int32.unpack(buffer) + 1;
            }
        }
    })
```

This could be used for other properties where we want to compute something dynamically.
Additionally, some way to process the fields object before the event is fired/before the buffer is sent (for example, added `process` and `unprocess` properties in some packet objects) could be useful for some packets where we want to do some extra processing on the packet before the application gets it.

**Comment by IvanSanchez** [comment]
> Additionally, some way to process the fields object before the event is fired/before the buffer is sent [...]

Yeah, that sounds exactly like middleware (like the stuff in `express.js` routers) and is not a completely bad idea. And this could be used to:
- Fix up the player ship index
- Cast/uncast some fields to `Bool`
- Cast/uncast some fields to `enum`

There should be a way to create the `asbs.Socket` specifying which of these middlewares are wanted. Another idea I had in my mind is to create different sets of datatypes, so you could specify "I want to use the basic datatypes", "I want to use the basic and enum datatypes" or "I want to use the basic and Bool datatypes".

Right now I'm not sure which is the best way to implement this.

**Comment by mrfishie** [comment]
Yeah, that would be useful for more complex packets, however I think we should ensure that most packets are able to be written without middleware (e.g. enums/bools as a type).
